dali/kernels/imgproc/convolution/convolution_gpu.h

```diff
@@ -0,0 +1,175 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_CONVOLUTION_CONVOLUTION_GPU_H_
+#define DALI_KERNELS_IMGPROC_CONVOLUTION_CONVOLUTION_GPU_H_
+
+#include "dali/core/boundary.h"
+#include "dali/core/convert.h"
+#include "dali/core/format.h"
+#include "dali/core/tensor_view.h"
+#include "dali/kernels/common/utils.h"
+#include "dali/kernels/imgproc/convolution/cutlass/device/gemm.h"
+#include "dali/kernels/kernel.h"
+#include "dali/pipeline/util/operator_impl_utils.h"
+
+namespace dali {
+namespace kernels {
+
+/**
+ * @brief Apply convolution with 1-channel `window` in specified axis.
+ *
+ * Operation is done by using GEMM and generating the matrix from convolution window on the fly
+ */
+template <typename Out, typename In, typename W, int ndim, int axis, bool has_channels = true>
+struct ConvolutionGpu {
+  KernelRequirements Setup(KernelContext& ctx, const TensorListShape<ndim>& in_shape,
+                           const TensorListShape<1>& window_size) {
+    KernelRequirements req;
+    ScratchpadEstimator se;
+    DALI_ENFORCE(
+        in_shape.size() == window_size.size(),
+        make_string("Provided input shape and window sizes do not mach in number of samples: ",
+                    in_shape.size(), " vs ", window_size.size(), "."));
+    int num_samples = in_shape.size();
+    for (int i = 0; i < num_samples; i++) {
+      int num_channels = has_channels ? in_shape[i][ndim - 1] : 1;
+      DALI_ENFORCE(window_size[i][0] % 2 == 1,
+                   make_string("Kernel window should have odd length, got: ", window_size,
+                               " for sample ", i, "."));
+
+      DALI_ENFORCE((window_size[i][0] / 2) * num_channels < kMaxRadiusSpan,
+                   make_string("Kernel window too big for sample ", i, "."));
+    }
+    se.add<W>(AllocType::Host, num_samples * kWindowCopyBufferSize);
+    se.add<W>(AllocType::GPU, num_samples * kWindowCopyBufferSize);
+    req.scratch_sizes = se.sizes;
+    req.output_shapes.push_back(in_shape);
+    return req;
+  }
+
+  void Run(KernelContext& ctx, const TensorListView<StorageGPU, Out, ndim> out,
+           const TensorListView<StorageGPU, const In, ndim>& in,
+           const TensorListView<StorageCPU, const W, 1>& window, W scale = 1) {
+    int num_samples = in.size();
+    auto* window_tmp_buffer_host =
+        ctx.scratchpad->Allocate<W>(AllocType::Host, num_samples * kWindowCopyBufferSize);
+    auto* window_tmp_buffer_gpu =
+        ctx.scratchpad->Allocate<W>(AllocType::GPU, num_samples * kWindowCopyBufferSize);
+
+    // Pad and align windows
+    AlignWindows(window_tmp_buffer_host, window, in.shape);
+    cudaMemcpyAsync(window_tmp_buffer_gpu, window_tmp_buffer_host,
```

Comment:
Where the `window_tmp_buffer_host` is populated? I was just created in L66.

Reply:
Looks like it's populated in `AlignWindows`.

Reply:
`// Pad and align windows` doesn't describe this, as well as the function name itself.

Reply:
Done